kernel/timeout.c

```diff
@@ -81,15 +89,40 @@ static s32_t next_timeout(void)
 	return ret;
 }
 
-void z_add_timeout(struct _timeout *to, _timeout_func_t fn, s32_t ticks)
+void z_add_timeout(struct _timeout *to, _timeout_func_t fn,
+		   k_timeout_t expires)
 {
+#ifdef CONFIG_SYS_TIMEOUT_LEGACY_API
+	k_ticks_t ticks = k_ms_to_ticks_ceil32(expires);
+#else
+	k_ticks_t ticks = expires.ticks;
+#endif
+
+	/* In the ISR, we know we're perfectly aligned to the start of
+	 * the current tick, so new timeouts can use exact tick
+	 * counts.  At arbitrary times, we need to wait for the next
+	 * tick to start counting so we don't expire too soon.
+	 */
+	if (!announcing) {
+		ticks += 1;
+	}
+
 	__ASSERT(!sys_dnode_is_linked(&to->node), "");
 	to->fn = fn;
 	ticks = MAX(1, ticks);
```

Comment:
Will this work with absolute timeouts coded now as negative value?

Reply:
I realized early this morning that the absolute timeout capability is not yet implemented.  I updated my review to request the API be removed for now.

Reply:
Fixed.  Yeah, this was adjusting ticks before checking if they were absolute or not.